src/libsystemd-network/sd-dhcp6-lease.c

```diff
@@ -473,6 +473,38 @@ int sd_dhcp6_lease_get_captive_portal(sd_dhcp6_lease *lease, const char **ret) {
         return 0;
 }
 
+int dhcp6_lease_insert_private_option(sd_dhcp6_lease *lease, uint8_t tag, const void *data, uint8_t len) {
+        struct sd_dhcp6_raw_option *option, *before = NULL;
+
+        assert(lease);
+
+        LIST_FOREACH(options, cur, lease->private_options) {
+                if (tag < cur->tag) {
+                        before = cur;
+                        continue;
+                }
+                if (tag == cur->tag) {
+                        log_debug("Ignoring duplicate option, tagged %i.", tag);
+                        return 0;
+                }
+        }
+
+        option = new(struct sd_dhcp6_raw_option, 1);
+        if (!option)
+                return -ENOMEM;
+
+        option->tag = tag;
+        option->length = len;
+        option->data = memdup(data, len);
+        if (!option->data) {
+                free(option);
```

Comment:
@yuwata : I needed a clarification. I had added _cleanup_ attribute for freeing the memory, but this is a failure case where we will free the memory of struct sd_dhcp6_raw_option which was allocated when memdup fails. Adding _cleanup_ here will free the memory once the variable goes out of scope even in success case. This would mean we will not be able to reference this memory while building the JSON. So we will free the memory here and use dhcp6_raw_option_free() in dhcp6_lease_free which will free all the memory that is allocated in the end.
Correct me if I am wrong.

Reply:
Fixed